fix(react-menu): Improving user experience by fixing a hoverDelay + click race condition on MenuTrigger

stevencomsft opened the pull request:
Fixing issue where menu item trigger click could have race condition with hover open and cause the menu to close immediately right as it is opening.

<!--
Thank you for submitting a pull request!

Please verify that:
* [ ] Code is up-to-date with the `master` branch
* [ ] Your changes are covered by tests (if possible)
* [ ] You've run `yarn change` locally


PR flow tips:
* [ ] Try to start with a Draft PR
* [ ] Once you're ready (ideally the pipeline is passing) promote your PR to Ready for Review. This step will auto-assign reviewers for your PR.
-->

## Previous Behavior

With a hover delay set on Menu (default is 500ms), a user can encounter (frequently) a race condition where they start to hover a menu trigger, which sets the internal open state of the menu to open without showing the menu as open visually for 500ms, so the user ends up clicking to try to open the menu, but the reaction + click takes just enough time that the menu blips open and then immediately closes due to the click event. 

## New Behavior

Prevent click-closing the menu while the hover delay is occurring. This allows users to click the menu trigger while it is visually closed and have the menu open as expected, even if an open event was already triggered by the hover action

stevencomsft commented:
@ling1726 I hear you have a lot of experience with Menu - was hoping to get your review/opinion on this change :)

ling1726 reviewed (changes requested):
Can you please provide some repro steps to this in the PR description?

I don't quite understand the root cause of the problem

> which sets the internal open state of the menu to open without showing the menu as open visually for 500ms

This statement is incorrect, when the user hovers over a trigger, the open state is not changed immediately, the update happens in a timeout  https://github.com/microsoft/fluentui/blob/84ccf12b6bbb147e40cdbca01bdf007ee59d2f25/packages/react-components/react-menu/src/components/Menu/useMenu.tsx#L213-L224

stevencomsft commented:
@ling1726  I made some changes after going back through the logic you quoted and missed that originally. updated the comments, description, and added some screen recordings :)
